docs/AQL/grammar/AqlParser.g4

```diff
@@ -0,0 +1,214 @@
+//
+//  description:  ANTLR4 parser grammar for Archetype Query Language (AQL)
+//  author:       Sebastian Iancu
+//  contributors: This version of the grammar is a complete rewrite of previously published antlr3 grammar,
+//                based on current AQL specifications in combination with other grammars from several AQL implementations.
+//                The openEHR Foundation would like to recognise the following people for their contributions:
+//                  - Chunlan Ma & Heath Frankel, Ocen Health Systems, Australia
+//                  - Bostjan Lah, Better, Slovenia
+//                  - Christian Chevalley, EHRBase, Germany
+//                  - Teun van Hemert & Michael Böckers, Nedap, Netherlands
+//  support:      openEHR Specifications PR tracker <https://specifications.openehr.org/releases/QUERY/open_issues>
+//  copyright:    Copyright (c) 2020 openEHR Foundation
+//  license:      Apache 2.0 License <http://www.apache.org/licenses/LICENSE-2.0.html>
+//
+
+parser grammar AqlParser;
+
+options { tokenVocab=AqlLexer; }
+
+selectQuery
+    : selectClause fromClause whereClause? orderByClause? limitClause? MINUSMINUS? EOF
+    ;
+
+selectClause
+    : SELECT top? selectExpr (COMMA selectExpr)*
+    ;
+
+fromClause
+    : FROM fromExpr
+    ;
+
+whereClause
+    : WHERE whereExpr
+    ;
+
+orderByClause
+    : ORDER BY orderByExpr (COMMA orderByExpr)*
+    ;
+
+limitClause
+    : LIMIT limit=NN_INTEGER (OFFSET offset=NN_INTEGER)?
+    ;
+
+
+selectExpr
+    : columnExpr (AS IDENTIFIER)?
+    ;
+
+fromExpr
+    : containsExpr
+    ;
+
+whereExpr
+    : NOT? identifiedExpr
+    | whereExpr AND whereExpr
+    | whereExpr OR whereExpr
+    | OPEN_PAR whereExpr CLOSE_PAR
+    ;
+
+orderByExpr
+    : identifiedPath order=(DESCENDING|DESC|ASCENDING|ASC)?
+    ;
+
+
+columnExpr
+    : identifiedPath
+    | aggregateFunctionCall
+    | functionCall
+    ;
+
+containsExpr
+    : classExprOperand (CONTAINS containsExpr)?
+    | containsExpr AND containsExpr
+    | containsExpr OR containsExpr
+    | OPEN_PAR containsExpr CLOSE_PAR
+    ;
+
+identifiedExpr
+    : EXISTS identifiedPath
+    | identifiedPath COMPARISON_OPERATOR identifiedOperand
+    | functionCall COMPARISON_OPERATOR identifiedOperand
+    | identifiedPath LIKE likeOperand
+    | identifiedPath MATCHES OPEN_CURLY matchesOperand CLOSE_CURLY
+    ;
+
+identifiedOperand
```

Comment:
no, intension was to capture only primitive values (e.g. "test", 1, 0.5, false) so that you can use it as an operand of COMPARISON_OPERATOR

Reply:
Should I still rename it as 'terminal' ?

Reply:
The word 'terminal' does have a particular meaning in the context of language grammars.There is a very formal definition for all this, but basically, terminal symbols in a grammar are symbols that only appear on the right-hand side of a production rule, and thus cannot be further reduced by any other production rule. They are basically the 'leaves' of a successful (e.g. error-free and complete) parse tree.

I am not sure if this is what @wolandscat means as well, I am not aware of a special meaning in OpenEHR which could very well be a lack of knowledge on my part, but I think this meaning of terminal might be a bit broad here, as there are a lot more 'terminals' in the language accepted by the AQL grammar. Maybe 'value' or 'literal' is appropriate?

Reply:
Ok, it's getting a bit clear. 
There is however no 'terminal' openEHR concept that plays a role here.
'Value' might be a bit better, but from my perspective not good enough. The 'identifiedOperand' is also not good, as it does not exists in specs - it is just my name-choice as it represents an operand in an identified expression, most likely on the right side of the operator (end-of-expression).
The same rule definition is also for function arguments.
We should find a better name ...

Reply:
Yep, 'terminal' is just a standard parser concept - it means an atom i.e. a leaf in the parser tree. 'Value' usually means 'literal value' in parser rules, and is useful for that purpose, and it is almost always one kind of terminal entity.

A general rule for parser grammars is to name things by the most general thing possible that applies. For example, if the RHS of an expression can be any kind of atomic entity, 'terminal' or a similar name makes sense, rather than 'xyzOperand'. Then you can reuse the 'terminal' in other places. You would use 'xyzOperand' only if the operand were some unique set of possibilities.